dap-dlv-go.el

```diff
@@ -0,0 +1,142 @@
+  (setq conf (pcase (plist-get conf :mode)
+               ("auto"
+		(dap-dlv-go--populate-auto-args conf))
+               ("debug"
+		(dap--put-if-absent
+		 conf :program (f-dirname (buffer-file-name))))
+               ("exec"
+		(dap--put-if-absent
+		 conf :program
+		 (f-expand (read-file-name "enter path to executable: "))))
+               ("remote"
+		(dap--put-if-absent conf :program (f-dirname (buffer-file-name)))
+		(dap--put-if-absent conf :host (read-string "enter host: " "127.0.0.1"))
+		(dap--put-if-absent conf :debugPort
+				    (string-to-number (read-string "enter port: " "2345"))))
+               ("local"
+                (dap--put-if-absent conf :cwd (f-dirname (buffer-file-name)))
+                (dap--put-if-absent
+		 conf :processId (string-to-number (read-string "enter pid: " "2345"))))))
+
+  (let ((debug-port (if (string= (plist-get conf :mode)
+				 "remote")
+			(plist-get conf :debugPort)
+		      (dap--find-available-port))))
+    (dap--put-if-absent conf :host "localhost")
+    (when (not (string= "remote" (plist-get conf :mode)))
+      (plist-put
+       conf :program-to-start
+       (format "%s dap --listen 127.0.0.1:%s --log" dap-dlv-go-delve-path debug-port)))
+    (plist-put conf :debugServer debug-port))
+
+  (if (stringp (plist-get conf :args)) (plist-put conf :args (split-string (plist-get conf :args))) ())
+  
+  (-> conf
+      (dap--put-if-absent :dlvToolPath dap-dlv-go-delve-path)
+      (dap--put-if-absent
+       :packagePathToGoModPathMap
+       (ht<-alist
+	`((,(f-dirname (buffer-file-name)) . ,(lsp-find-session-folder (lsp-session) (buffer-file-name))))))
+      
+      (dap--put-if-absent :type "go")
+      (dap--put-if-absent :name "Go Dlv Debug")))
+
+(defun dap-dlv-go--populate-auto-args (conf)
+  "Populate auto arguments according to CONF."
+  (dap--put-if-absent conf :program (buffer-file-name))
+
+  (if (string-suffix-p "_test.go" (buffer-file-name))
+      (plist-put conf :mode "test")
+    (plist-put conf :mode "debug")))
+
+(dap-register-debug-provider "go" 'dap-dlv-go--populate-default-args)
+
+(dap-register-debug-template "Go Dlv Launch File Configuration"
+                             (list :type "go"
+                                   :request "launch"
+                                   :name "Launch File"
+                                   :mode "auto"
+                                   :program nil
+                                   :buildFlags nil
+                                   :args nil
+                                   :env nil
+                                   :envFile nil))
+
+(dap-register-debug-template "Go Dlv Attach Configuration"
+                             (list :type "go"
+                                   :request "attach"
+                                   :name "Attach to running process"
+                                   :mode "auto"
+                                   :program nil
+                                   :buildFlags nil
```

Comment:
Do you have to include nils for flags that will never apply to the mode? (e.g. if you attach to running process, there is nothing to build)

Reply:
jep, we should drop them so they will not be passed to server.
here I took them out. so only args, env and envFile gets passed to server:
`
2022-04-26T20:29:28+03:00 debug layer=dap [<- from client]{"seq":2,"type":"request","command":"attach","arguments":{"type":"go","request":"attach","name":"Dlv Remote Debug\u003c8\u003e","mode":"remote","host":"100.115.92.202","args":null,"env":null,"envFile":null,"port":12345,"debugServer":12345,"dlvToolPath":"/home/apmattil/go/bin/dlv","packagePathToGoModPathMap":{"/home/apmattil/src/ari":"/home/apmattil/src/ari"}}}
`

Reply:
`envFile` is folded into `env` in vscode and delve only understands `env`. This is only passed to launched programs. Same with args - these are for when the program is started. You can find all the attach attributes that dlv recognizes here: https://pkg.go.dev/github.com/go-delve/delve/service/dap#AttachConfig. And launch here: https://pkg.go.dev/github.com/go-delve/delve/service/dap#LaunchConfig. Anything extra, dlv will ignore.